File: backend/job-application-worker/src/utils/logger.js
/**
 * Structured logging utility for Job Application Worker.
 * Provides JSON logging with service name and trace_id support.
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Context for trace_id (simple in-memory for now, can be enhanced with AsyncLocalStorage)
let traceId = null;

// File logging setup
let logFileStream = null;
const env = process.env.ENV || 'development';
const isProduction = env === 'production' || env === 'prod';
const logToFile = process.env.LOG_TO_FILE === 'true';
const logDir = process.env.LOG_DIR || 'logs';

// Initialize file logging if enabled (development only)
if (!isProduction && logToFile) {
  try {
    // Create log directory if it doesn't exist
    const logDirPath = path.resolve(process.cwd(), logDir);
    if (!fs.existsSync(logDirPath)) {
      fs.mkdirSync(logDirPath, { recursive: true });
    }
    
    const logFile = path.join(logDirPath, 'job-application-worker.log');
    logFileStream = fs.createWriteStream(logFile, { flags: 'a' });
  } catch (err) {
    console.error('Failed to create log file, falling back to stdout only', err);
  }
}

/**
 * Set trace_id for distributed tracing
 */
export const setTraceId = (id) => {
  traceId = id;
};

/**
 * Get trace_id from context
 */
export const getTraceId = () => {
  return traceId;
};

/**
 * Write log entry to both stdout and file (if enabled)
 */
const writeLog = (level, entry) => {
  const logLine = JSON.stringify(entry) + '\n';
  
  // Always write to stdout
  if (level === 'error') {
    console.error(logLine);
  } else if (level === 'warn') {
    console.warn(logLine);
  } else if (level === 'debug') {
    console.debug(logLine);
  } else {
    console.log(logLine);
  }
  
  // Also write to file if enabled
  if (logFileStream) {
    logFileStream.write(logLine);
  }
};

/**
 * Create a log entry with standard fields
 */
const createLogEntry = (level, message, meta = {}) => {
  const entry = {
    timestamp: new Date().toISOString(),
    level,
    service: 'job-application-worker',
    message,
    ...meta,
  };
  
  // Add trace_id if available
  if (traceId) {
    entry.trace_id = traceId;
  }
  
  return entry;
};

export const logger = {
  info: (message, meta = {}) => {
    const entry = createLogEntry('info', message, meta);
    writeLog('info', entry);
  },

  warn: (message, meta = {}) => {
    const entry = createLogEntry('warn', message, meta);
    writeLog('warn', entry);
  },

  error: (message, meta = {}) => {
    const entry = createLogEntry('error', message, meta);
    writeLog('error', entry);
  },
  
  debug: (message, meta = {}) => {
    if (!isProduction) {
      const entry = createLogEntry('debug', message, meta);
      writeLog('debug', entry);
    }
  },
  
  /**
   * Close log file stream (call on shutdown)
   */
  close: () => {
    if (logFileStream) {
      logFileStream.end();
      logFileStream = null;
    }
  },
};

File: backend/docs-service/app/logger.py
"""
Structured logging utility for Docs Service.
Provides JSON logging in production, text logging in development.
"""
import os
import sys
import structlog
from typing import Optional
from contextvars import ContextVar

# Context variable for trace_id (thread-safe)
trace_id_var: ContextVar[Optional[str]] = ContextVar("trace_id", default=None)


def configure_logging(env: str = "development", log_to_file: bool = False, log_dir: str = "logs"):
    """
    Configure structured logging for the service.
    
    Args:
        env: Environment name (development, production, etc.)
        log_to_file: Whether to write logs to files (development only)
        log_dir: Directory for log files (if log_to_file is True)
    """
    is_production = env.lower() in ("production", "prod")
    
    # Configure processors
    processors = [
        structlog.contextvars.merge_contextvars,  # Merge context variables
        structlog.stdlib.add_log_level,            # Add log level
        structlog.stdlib.add_logger_name,         # Add logger name
        structlog.processors.TimeStamper(fmt="iso"),  # ISO 8601 timestamp
        structlog.processors.StackInfoRenderer(),     # Stack traces
        structlog.processors.format_exc_info,          # Exception formatting
    ]
    
    if is_production:
        # JSON output for production
        processors.append(structlog.processors.JSONRenderer())
        log_level = "INFO"
    else:
        # Human-readable output for development
        processors.append(structlog.dev.ConsoleRenderer())
        log_level = "DEBUG"
    
    # Configure structlog
    structlog.configure(
        processors=processors,
        wrapper_class=structlog.stdlib.BoundLogger,
        context_class=dict,
        logger_factory=structlog.stdlib.LoggerFactory(),
        cache_logger_on_first_use=True,
    )
    
    # Configure standard library logging
    import logging
    logging.basicConfig(
        format="%(message)s",
        stream=sys.stdout,
        level=getattr(logging, log_level.upper()),
    )
    
    # If file logging is enabled in development, set up file handler
    if not is_production and log_to_file:
        os.makedirs(log_dir, exist_ok=True)
        log_file = os.path.join(log_dir, "docs-service.log")
        file_handler = logging.FileHandler(log_file)
        file_handler.setFormatter(logging.Formatter("%(message)s"))
        
        root_logger = logging.getLogger()
        root_logger.addHandler(file_handler)


def get_logger(name: str = "woragis.docs-service"):
    """
    Get a structured logger instance.
    
    Args:
        name: Logger name
        
    Returns:
        Bound logger with service name and trace_id support
    """
    logger = structlog.get_logger(name)
    
    # Add service name to all logs
    logger = logger.bind(service="docs-service")
    
    # Add trace_id from context if available
    trace_id = trace_id_var.get()
    if trace_id:
        logger = logger.bind(trace_id=trace_id)
    
    return logger


def set_trace_id(trace_id: str):
    """Set trace_id in context for distributed tracing."""
    trace_id_var.set(trace_id)


def get_trace_id() -> Optional[str]:
    """Get trace_id from context."""
    return trace_id_var.get()
